3scale/3scale-operator
# THREESCALE-8999-Allow adding labels to components via APIManager CR

## What
Allow adding labels to components via APIManager CR
Jira: [THREESCALE-8999](https://issues.redhat.com/browse/THREESCALE-8999)

Adding optional Labels definition for the following 3scale pods:

Component | 
-- |
apicast-production |
apicast-staging |
backend-cron |
backend-listener |
backend-worker |
backend-redis |
system-app |
system-sidekiq |
system-searchd |
system-memcache |
system-mysql |
system-postgresql |
system-redis |
zync |
zync-database |
zync-que |

## Verification

1. Check out this branch
2. Run 3scale-operator
```
make install
oc new-project 3scale
make run
```
3. Create s3 secret
```
oc apply -f - <<EOF                                                             
---             
apiVersion: v1                      
kind: Secret    
metadata:
  creationTimestamp: null
  name: aws-auth
stringData:                                               
  AWS_ACCESS_KEY_ID: something      
  AWS_SECRET_ACCESS_KEY: something
  AWS_BUCKET: something
  AWS_REGION: us-east-1    
type: Opaque                   
EOF
```

4. Create APIManager
```
oc apply -f  - <<EOF
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager1
spec:
  system:
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: aws-auth
    appSpec:
      labels:
        systemAppL1: systemAppL1
        systemAppL2: systemAppL2
    searchdSpec:
      labels:
        systemSearchdL1: systemSearchdL1
        systemSearchdL2: systemSearchdL2
    sidekiqSpec:
      labels:
        systemSidekiqL1: systemSidekiqL1
        systemSidekiqL2: systemSidekiqL2
    memcachedLabels:
      systemMemCachedL1: systemMemCachedL1
      systemMemCachedL2: systemMemCachedL2
    redisLabels:
      systemRedisL1: systemRedisL1
      systemRedisL2: systemRedisL2
    database:
      postgresql:
        labels:
          systemDBPostgresL1: systemDBPostgresL1
          systemDBPostgresL2: systemDBPostgresL2
  zync:
    appSpec:
      labels:
        zyncAppL1: zyncAppL1
        zyncAppL2: zyncAppL2
    queSpec:
      labels:
        zyncQueL1: zyncQueL1
        zyncQueL2: zyncQueL2
    databaseLabels:
      zyncDBL1: zyncDBL1
      zyncDBL2: zyncDBL2
  backend:
    cronSpec:
      labels:
        backendCronL1: backendCronL1
        backendCronL2: backendCronL2
    listenerSpec:
      labels:
        backendListenerL1: backendListenerL1
        backendListenerL2: backendListenerL2
    workerSpec:
      labels:
        backendWorkerL1: backendWorkerL1
        backendWorkerL2: backendWorkerL2
    redisLabels:
      backendRedisL1: backendRedisL1
      backendRedisL2: backendRedisL2
  apicast:
    productionSpec:
      labels:
        apicastProdL1: apicastProdL1
        apicastProdL2: apicastProdL2
    stagingSpec:
      labels:
        apicastStgL1: apicastStgL1
        apicastStgL2: apicastStgL2
  wildcardDomain: apps.some-valid-domain

EOF
```

5. Check that the deploymentConfigs contain the labels:
```
 kubectl get deploymentconfigs -n 3scale -o custom-columns="DEPLOYMENTCONFIG:metadata.name, LABELS:spec.template.metadata.labels"
```
7. Update the labels and check that the label has been updated in the deploymentConfig.

9. Check that deleting a label in the APIManager removes the label from the deploymentConfig.

11. Check that creating a new label in the APIManager creates a new label in the deploymentConfig.
